### PR TITLE
add sda-download to helm charts

### PR DESCRIPTION
**Pull request by blankdots**
this will add sda-download to helm charts (ultimately with the purpose of replace sda-doa)
depends on the changes in https://github.com/neicnordic/sda-download/pull/76
depends on the changes in https://github.com/neicnordic/sda-download/pull/77
depends on the changes in https://github.com/neicnordic/sda-download/pull/78
depends on the changes in https://github.com/neicnordic/sda-download/pull/82

This is not a breaking change.
Changes made:
* `sda-svc` Chart version bumped to v0.16.0
* added `sda-download` to the charts with the option to start in posix and s3
* added tests for the `sda-download` endpoint
* clean up of some python dependencies
* s3 bakend configuration enabled
* https enabled

### CODE REVIEW

**Review by norling — APPROVED** [review]
I'm no kubernetes expert, but apart from a pair of typos this looks good to me!

**Comment by blankdots** [comment]
this will be merged after #104 is merged

File: charts/sda-svc/templates/doa-deploy.yaml [hunk 1]
```diff
@@ -146,7 +146,7 @@ spec:
         - name: CRYPT4GH_PRIVATE_KEY_PASSWORD_PATH
           value: "{{ template "c4ghPath" . }}/passphrase"
         - name: OPENID_CONFIGURATION_URL
-          value: "{{ .Values.global.elixir.oidcdHost }}.well-known/openid-configuration"
+          value: "{{ .Values.global.elixir.oidcdHost }}/.well-known/openid-configuration"
```
Comment (jbygdell) [inline charts/sda-svc/templates/doa-deploy.yaml:149]: This change to DOA should probably not be part of this PR

File: charts/sda-svc/templates/download-ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+{{- if .Values.global.ingress.labels }}
+{{ toYaml .Values.global.ingress.labels | indent 4 }}
+{{- end }}
+  annotations:
+    kubernetes.io/ingress.class: "nginx"
+    nginx.ingress.kubernetes.io/rewrite-target: "/"
+    nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
```
Comment (jbygdell) [inline charts/sda-svc/templates/download-ingress.yaml:18]: This needs to be modular for a non TLS setup

File: charts/sda-svc/templates/download-secrets.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{{- if not .Values.global.pkiService }}
+---
+{{- $download := .Files.Glob "files/download*" }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ template "sda.fullname" . }}-download-certs
+data:
+{{ ( .Files.Glob "files/ca.crt" ).AsSecrets | trim | indent 2 }}
+{{ ( $download ).AsSecrets | indent 2 }}
```
Comment (jbygdell) [inline charts/sda-svc/templates/download-secrets.yaml:30]:
```suggestion
```

Comment (jbygdell) [reply charts/sda-svc/templates/download-secrets.yaml:30]: The secret for the certificates needs to be created manually as of #104

File: charts/sda-svc/values.yaml [hunk 4]
```diff
@@ -184,6 +185,22 @@ global:
     servicePort: 443
     envFile: env
 
+  download:
+    useTLS: true
```
Comment (jbygdell) [inline charts/sda-svc/values.yaml:189]:
```suggestion
```

Comment (jbygdell) [reply charts/sda-svc/values.yaml:189]: This should be handled by the global tls configuation `global.tls.enabled`

File: dev_tools/scripts/make-certs.sh [hunk 5]
```diff
@@ -49,6 +49,10 @@ openssl x509 -req -in "./${basedir}/verify.csr" -days "${days}" -CA "./${basedir
 openssl req -config "$(dirname "$0")"/ssl.cnf -new -nodes -newkey rsa:4096 -keyout "./${basedir}/doa.key" -out "./${basedir}/doa.csr" -extensions doa_cert
 openssl x509 -req -in "./${basedir}/doa.csr" -days "${days}" -CA "./${basedir}/ca.crt" -CAkey "./${basedir}/ca.key" -set_serial 01 -out "./${basedir}/doa.crt" -extensions doa_cert -extfile "$(dirname "$0")"/ssl.cnf
 
+# Create certificate for download
+openssl req -config "$(dirname "$0")"/ssl.cnf -new -nodes -newkey rsa:4096 -keyout "./${basedir}/download.key" -out "./${basedir}/download.csr" -extensions download_cert
```
Comment (jbygdell) [inline dev_tools/scripts/make-certs.sh:53]:
```suggestion
openssl req -config "$(dirname "$0")"/ssl.cnf -new -nodes -newkey rsa:4096 -keyout "./${basedir}/download.key" -out "./${basedir}/download.csr" -extensions download_cert -subj "/CN=lega_out/CN=admin/"
```

Comment (jbygdell) [reply dev_tools/scripts/make-certs.sh:53]: CN is needed for peer verification to function properly

File: charts/sda-svc/values.yaml [hunk 6]
```diff
@@ -318,6 +339,24 @@ doa:
...
+      memory: "256Mi"
+      cpu: "100m"
+    limits:
+      memory: "512Mi"
+      cpu: "1000m"
+# Extra annotations to attach to the service pods
+# This should be a multi-line string mapping directly to the a map of
+# the annotations to apply to the service pods
+  annotations: {}
+
```
Comment (jbygdell) [inline charts/sda-svc/values.yaml:359]:
```suggestion
  annotations: {}
  tls:
    secretName: ""
```

File: charts/sda-svc/templates/download-deploy.yaml [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      {{- end }}
+        ports:
+        - name: download
+          containerPort: {{ ternary 8080 8443 ( empty .Values.global.download.useTLS ) }}
+          protocol: TCP
+        livenessProbe:
+          httpGet:
+            port: download
+            path: /health
+            scheme: {{ ternary "HTTP" "HTTPS" ( empty .Values.global.download.useTLS ) }}
```
Comment (jbygdell) [inline charts/sda-svc/templates/download-deploy.yaml:171]:
```suggestion
            scheme: {{ ternary "HTTPS" "HTTP" .Values.global.tls.enabled }}
```

File: charts/sda-svc/templates/download-deploy.yaml [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+            httpHeaders:
+            - name: Host
+              value: {{ .Values.global.ingress.hostName.download }}
+          initialDelaySeconds: 20
+          periodSeconds: 10
+        readinessProbe:
+          httpGet:
+            port: download
+            path: /health
+            scheme: {{ ternary "HTTP" "HTTPS" ( empty .Values.global.download.useTLS ) }}
```
Comment (jbygdell) [inline charts/sda-svc/templates/download-deploy.yaml:181]:
```suggestion
            scheme: {{ ternary "HTTPS" "HTTP" .Values.global.tls.enabled }}
```

File: charts/sda-svc/templates/download-deploy.yaml [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+          value: {{ template "c4ghPath" . }}/{{ default "c4gh.sec.pem" .Values.global.c4gh.file }}
+      {{- if or (eq "verify-ca" .Values.global.db.sslMode) (eq "verify-full" .Values.global.db.sslMode) }}
+        - name: DB_CACERT
+          value: {{ include "tlsPath" . }}/ca.crt
+      {{- end }}
+      {{- if eq "verify-full" .Values.global.db.sslMode }}
+        - name: DB_CLIENTCERT
+          value: {{ include "tlsPath" . }}/download.crt
+        - name: DB_CLIENTKEY
+          value: {{ include "tlsPath" . }}/download.key
```
Comment (jbygdell) [inline charts/sda-svc/templates/download-deploy.yaml:105]:
```suggestion
        - name: DB_CLIENTCERT
          value: {{ include "tlsPath" . }}/tls.crt
        - name: DB_CLIENTKEY
          value: {{ include "tlsPath" . }}/tls.key
```

Comment (jbygdell) [reply charts/sda-svc/templates/download-deploy.yaml:105]: This is what it will look like after #110

File: charts/sda-svc/templates/download-deploy.yaml [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        {{- if eq "posix" .Values.global.archive.storageType }}
+        - name: archive
+          mountPath: {{ .Values.global.archive.volumePath | quote }}
+        {{- end }}
+      volumes:
+      {{- if not .Values.global.pkiService }}
+        - name: {{ ternary "tls" "tls-certs" (empty .Values.global.pkiPermissions) }}
+          secret:
+            defaultMode: 0440
+            secretName: {{ template "sda.fullname" . }}-download-certs
```
Comment (jbygdell) [inline charts/sda-svc/templates/download-deploy.yaml:209]: This will need to be a projected volume as of #104

**Review by jbygdell — DISMISSED** [review]
Some stuff, mostly hints on what #104 and #110 will bring

File: charts/sda-svc/templates/doa-deploy.yaml (hunk 1 above)
Comment (blankdots) [reply charts/sda-svc/templates/doa-deploy.yaml:149]: ok, will remove

**Comment by blankdots** [comment]
will try to address the rest of the comments, as batch, will see if I can rebase on top of #110

**Review by jbygdell — APPROVED** [review]

**Comment by blankdots** [comment]
Will merge this, as this has been open for some time and all issues addressed.
